src/primitives/fieldvec.rs

```diff
@@ -0,0 +1,380 @@
+    #[inline]
+    pub fn has_data(&self) -> bool { self.len <= NO_ALLOC_MAX_LENGTH || cfg!(feature = "alloc") }
+
+    /// Panics if [`Self::has_data`] is false, with an informative panic message.
+    #[inline]
+    pub fn assert_has_data(&self) {
+        assert!(
+            self.has_data(),
+            "checksums of {} characters (more than {}) require the `alloc` feature of `bech32` to be enabled",
+            NO_ALLOC_MAX_LENGTH,
+            self.len,
+        );
+    }
+
+    /// Number of stored field elements
+    #[inline]
+    pub fn len(&self) -> usize { self.len }
+
+    /// Whether the vector is empty
+    #[inline]
+    pub fn is_empty(&self) -> bool { self.len == 0 }
+
+    /// Returns an immutable iterator over the elements in the vector.
+    ///
+    /// # Panics
+    ///
+    /// Panics if [`Self::has_data`] is false.
+    pub fn iter(&self) -> slice::Iter<F> {
+        if self.len > NO_ALLOC_MAX_LENGTH {
+            self.assert_has_data();
+            #[cfg(feature = "alloc")]
+            return self.inner_v[..self.len].iter();
+        }
+        self.inner_a[..self.len].iter()
+    }
+
+    /// Returns a mutable iterator over the elements in the vector.
+    ///
+    /// # Panics
+    ///
+    /// Panics if [`Self::has_data`] is false.
+    pub fn iter_mut(&mut self) -> slice::IterMut<F> {
+        if self.len > NO_ALLOC_MAX_LENGTH {
+            self.assert_has_data();
+            #[cfg(feature = "alloc")]
+            return self.inner_v[..self.len].iter_mut();
+        }
+        self.inner_a[..self.len].iter_mut()
+    }
+}
+
+impl<F: Field> FieldVec<F> {
+    /// Constructor from the powers of an element, from 0 upward.
+    #[inline]
+    pub fn from_powers(elem: F, n: usize) -> Self {
+        iter::successors(Some(F::ONE), |gen| Some(elem.clone() * gen)).take(n + 1).collect()
+    }
+
+    /// Multiply the elements of two vectors together, pointwise.
+    #[inline]
+    pub fn mul_assign_pointwise(&mut self, other: &Self) {
+        assert_eq!(self.len, other.len, "cannot add vectors of different lengths");
+        for (i, fe) in self.iter_mut().enumerate() {
+            *fe *= &other[i];
+        }
+    }
+
+    /// Multiply the elements of two vectors together, pointwise.
+    pub fn mul_pointwise(mut self, other: &Self) -> Self {
+        self.mul_assign_pointwise(other);
+        self
+    }
+
+    #[inline]
+    /// Lifts a vector of field elements to a vector of elements in an extension
+    /// field, via the inclusion map.
+    pub fn lift<E: Field + From<F>>(&self) -> FieldVec<E> {
+        self.iter().cloned().map(E::from).collect()
+    }
+}
+
+impl<F: Default> FieldVec<F> {
+    /// Pushes an item onto the end of the vector.
+    pub fn push(&mut self, item: F) {
+        self.len += 1;
+        self.assert_has_data();
+
+        #[cfg(not(feature = "alloc"))]
+        {
+            self.inner_a[self.len - 1] = item;
+        }
```

Comment:
```suggestion
        #[cfg(not(feature = "alloc"))]
        self.inner_a[self.len - 1] = item;
```

Or `self.inner_a[self.len] = item` if you use my other suggestion.

Reply:
Did you try compiling this suggestion?

Reply:
> error[E0658]: attributes on expressions are experimental

Worst review suggestion ever :)